skbio/alignment/_tabular_msa.py

```diff
@@ -690,12 +674,124 @@ def reindex(self, key=None, keys=None):
+        key : hashable, optional
+            Key for the MSA to use for the appended sequence. Note that
+            `key` cannot be combined with `minter`.
+
+        Raises
+        ------
+        OperationError
+            If both key and minter are provided.
+        OperationError
+            If MSA has keys but no key or minter was provided for the sequence
+            being appended.
+        OperationError
+            If key was provided but the MSA does not have keys.
+        OperationError
+            If minter was provided but the MSA does not have keys.
+        TypeError
+            If the sequence object is a type that doesn't have an alphabet
+        TypeError
+            If the type of the sequence does not match the dtype of the MSA.
+        ValueError
+            If the length of the sequence does not match the number of
+            positions in the MSA.
+
+        See Also
+        --------
+        reindex
+
+        Notes
+        -----
+        The MSA is not automatically re-aligned when a sequence is appended.
+        Therefore, this operation is not necessarily meaningful on its own.
+
+        Examples
+        --------
+        >>> from skbio import DNA, TabularMSA
+        >>> msa = TabularMSA([DNA('')])
+        >>> msa.append(DNA(''))
+        >>> msa == TabularMSA([DNA(''), DNA('')])
+        True
+
+        >>> msa = TabularMSA([DNA('', metadata={'id': 'a'})], minter='id')
+        >>> msa.append(DNA('', metadata={'id': 'b'}), minter='id')
+        >>> msa == TabularMSA([DNA('', metadata={'id': 'a'}),
+        ...                    DNA('', metadata={'id': 'b'})], minter='id')
+        True
+        """
+
+        if key is not None and minter is not None:
+            raise ValueError(
+                "Cannot use both `minter` and `key` at the same time.")
+
+        new_key = None
+        if self.has_keys():
+            if key is None and minter is None:
+                raise OperationError(
+                    "MSA has keys but no key or minter was provided.")
+            elif key is not None:
+                new_key = key
+            elif minter is not None:
+                new_key = resolve_key(sequence, minter)
+        else:
+            if key is not None:
+                raise OperationError(
+                    "key was provided but MSA does not have keys.")
+            elif minter is not None:
+                raise OperationError(
+                    "minter was provided but MSA does not have keys.")
+
+        self._add_sequence(sequence, new_key)
+
+    def _add_sequence(self, sequence, key=None):
+        msa_is_empty = not len(self)
+        dtype = type(sequence)
+        if msa_is_empty:
+            if not issubclass(dtype, IUPACSequence):
+                raise TypeError(
+                    "`sequence` must be a scikit-bio sequence object "
+                    "that has an alphabet, not type %r" % dtype.__name__)
+            self._dtype = dtype
```

Comment:
This block has the transactional update problem (the block below looks great though!). Can you fix this and add a unit test? Then this PR is ready to go!